src/platform/windows/misc.cpp

```diff
@@ -550,6 +556,243 @@ namespace platf {
+        return converter.from_bytes(raw_cmd);
+      }
+
+      // If the target is a URL, the class is found using the URL scheme (prior to and not including the ':')
+      lookup_string = scheme.data();
+    }
+    else {
+      // If the target is not a URL, assume it's a regular file path
+      auto extension = PathFindExtensionW(raw_target.c_str());
+      if (extension == nullptr || *extension == 0) {
+        // If the file has no extension, assume it's a command and allow CreateProcess()
+        // to try to find it via PATH
+        return converter.from_bytes(raw_cmd);
+      }
+
+      // For regular files, the class is found using the file extension (including the dot)
+      lookup_string = extension;
+    }
+
+    // Override HKEY_CLASSES_ROOT and HKEY_CURRENT_USER to ensure we query the correct class info
+    if (!override_per_user_predefined_keys(token)) {
+      return converter.from_bytes(raw_cmd);
+    }
+
+    // Find the command string for the specified class
+    std::array<WCHAR, MAX_PATH> shell_command_string;
+    DWORD out_len = shell_command_string.size();
+    res = AssocQueryStringW(ASSOCF_NOTRUNCATE, ASSOCSTR_COMMAND, lookup_string.c_str(), L"open", shell_command_string.data(), &out_len);
+
+    // In some cases (UWP apps), we might not have a command for this target. If that happens,
+    // we'll have to launch via cmd.exe. This prevents proper job tracking, but that was already
+    // broken for UWP apps anyway due to how they are started by Windows. Even 'start /wait'
+    // doesn't work properly for UWP, so really no termination tracking seems to work at all.
+    //
+    // FIXME: Maybe we can improve this in the future.
+    if (res == HRESULT_FROM_WIN32(ERROR_NO_ASSOCIATION)) {
+      BOOST_LOG(warning) << "Using trampoline to handle target: "sv << raw_cmd;
+      std::wcscpy(shell_command_string.data(), L"cmd.exe /c start \"\" \"%1\" %*");
+      res = S_OK;
+    }
+
+    // Reset per-user keys back to the original value
+    override_per_user_predefined_keys(NULL);
+
+    if (res != S_OK) {
+      BOOST_LOG(warning) << "Failed to query command string for raw command: "sv << raw_cmd << " ["sv << util::hex(res).to_string_view() << ']';
+      return converter.from_bytes(raw_cmd);
+    }
+
+    // Finally, construct the real command string that will be passed into CreateProcess().
+    // We support common substitutions (%*, %1, %2, %L, %W, %V, etc), but there are other
+    // uncommon ones that are unsupported here.
+    //
+    // https://web.archive.org/web/20111002101214/http://msdn.microsoft.com/en-us/library/windows/desktop/cc144101(v=vs.85).aspx
+    std::wstring cmd_string { shell_command_string.data() };
+    size_t match_pos = 0;
+    while ((match_pos = cmd_string.find_first_of(L'%', match_pos)) != std::wstring::npos) {
+      std::wstring match_replacement;
+      auto next_char = std::tolower(match_pos + 1 < cmd_string.size() ? cmd_string.at(match_pos + 1) : 0);
+      switch (next_char) {
```

Comment:
I don't know how to do it in C++, but I'd think this would be a good case to use regex for the different substitutions.

https://regex101.com/r/LuFITp/2

Reply:
Due to the replacement rules and the simplicity of the the way those argument substitutions work, I don't think a regex here would really simplify things.

`std::regex_replace()` can't do multiple different replacements in a single call. Calling `std::regex_replace()` multiple times would be worse than the single iteration through the string that the code here does, and it would incorrectly handle replacements where the replacement itself contains another `%1` or whatever.

We could use `std::regex_search()` and process the matches in a loop, but that basically ends up looking like the loop we have here.